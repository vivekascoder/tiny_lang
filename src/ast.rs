// Types for our language.

use std::{fmt::Display, rc::Rc};

pub type Program = Vec<Statement>;
pub type BlockStatement = Vec<Statement>;

// Precedence order
#[derive(PartialEq, PartialOrd, Debug, Clone)]
pub enum Precedence {
    Lowest,
    Equals,      // ==
    LessGreater, // > or <
    Sum,         // +
    Product,     // *
    Prefix,      // -X or !X
    Call,        // myFunction(x)
    Index,       // array[index]
    LRShift,
    BitwiseAnd,
    BitwiseXor,
    BitwiseOr,
}

#[derive(Clone, Debug, PartialEq)]
pub enum Expr {
    Ident(Ident),
    Infix(Infix, Box<Expr>, Box<Expr>),
    Prefix(Prefix, Box<Expr>),
    Literal(Literal),
    Call(FunctionCall),
}

#[derive(Clone, Debug, PartialEq)]
pub struct FunctionCall {
    pub parameters: Vec<Expr>,
    pub name: Rc<str>,
}

#[derive(Clone, Debug, PartialEq)]
pub struct NativeFunctionCall {
    pub parameters: Vec<ExprResult>,
    pub name: Rc<str>,
}

#[derive(Clone, Debug, PartialEq)]
pub enum ExprResult {
    Bool(bool),
    UnsignedInteger(usize),
    SignedInteger(isize),
    Char(char),
    Return(Box<ExprResult>),
    Void,
}

impl Display for ExprResult {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Bool(b) => write!(f, "{}", b),
            Self::UnsignedInteger(i) => write!(f, "{}", i),
            Self::SignedInteger(i) => write!(f, "{}", i),
            Self::Void => write!(f, ""),
            Self::Return(v) => write!(f, "{}", *v),
            Self::Char(c) => write!(f, "{}", c),
        }
    }
}

#[derive(Clone, Debug, PartialEq)]
pub enum MemoryObject {
    ExprResult(ExprResult),
    Function(Function),
}

#[derive(Clone, Debug, PartialEq)]
pub enum Infix {
    Plus,
    Minus,
    Divide,
    Multiply,
    DoubleEqual,
    NotEqual,
    GreaterThanEqual,
    GreaterThan,
    LessThanEqual,
    LessThan,

    // Bitwise operations
    LeftShift,
    RightShift,
    BitwiseAnd,
    BitwiseXor,
    BitwiseOr,
}

#[derive(PartialEq, Clone, Debug)]
pub enum Prefix {
    Plus,
    Minus,
    Not,
}

#[derive(PartialEq, Clone, Debug)]
pub enum Literal {
    UnsignedInteger(usize),
    Bool(bool),
    Char(char),
    String(Rc<str>),
}

// enum TVec {
//     LiteralVec<Literal>
// }

// WASM specific choices
// pub enum Literal {
//     I32(i32),
//     I64(i64),
//     F32(i32),
//     F64(i64),
//     Bool(bool),
//     Char(char),
// }

#[derive(Clone, Debug, PartialEq)]
pub struct Ident(pub Rc<str>);

#[derive(Clone, Debug, PartialEq)]
pub enum Statement {
    Let(Ident, Option<Type>, Expr),
    Mutate(Ident, Expr),
    Function(Function),
    Return(Expr),
    If(Condition),
    Expr(Expr),
    While(While),
}

#[derive(Clone, Debug, PartialEq)]
pub struct While {
    pub condition: Expr,
    pub body: BlockStatement,
}

#[derive(Clone, Debug, PartialEq)]
pub struct Condition {
    pub condition: Expr,
    pub if_body: BlockStatement,
    pub else_body: Option<BlockStatement>,
}

#[derive(Clone, Debug, PartialEq, Copy)]
pub enum Type {
    UnsignedInteger,
    SignedInteger,
    Bool,
    Char,
    String,
}

#[derive(Clone, Debug, PartialEq)]
pub struct Function {
    pub name: Rc<str>,
    pub params: Vec<(Ident, Type)>,
    pub return_type: Option<Type>,
    pub body: Vec<Statement>,
}

#[derive(Debug, Clone, PartialEq)]
pub enum TokenType {
    EOF,
    Identifier(Rc<str>),
    Usize(usize),
    Boolean(bool),
    String(Rc<str>),
    Colon,
    SemiColon,

    // Operators
    Equal,
    DoubleEqual,
    Plus,
    Minus,
    Multiply,
    Divide,
    Mod,
    NotEqual,
    LessThan,
    LessThanEqual,
    GreaterThan,
    GreaterThanEqual,
    Bang,
    LParen,
    RParen,
    LBrace, // {
    RBrace, // }
    Comma,
    SQuote,     // '
    DQuote,     // "
    BSlash,     // \
    NewLine,    // \n
    Tab,        // \t
    Space,      // ' '
    LeftShift,  // <<
    RightShift, // >>
    Ampersand,  // &
    Carrot,     // ^
    Pipe,       // |

    // Keywords
    KeywordLet,
    KeywordUsize,
    KeywordIsize,
    KeywordBool,
    KeywordFun,
    KeywordReturn,
    KeywordIf,
    KeywordElse,
    KeywordVoid,
    KeywordChar,
    KeywordWhile,
    KeywordStr,

    // Symbol
    SymbolReturn,
}

/// implement the Debug trait for all the variants of TokenType
impl Display for TokenType {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        match self {
            TokenType::EOF => write!(f, "EOF"),
            TokenType::Identifier(s) => write!(f, "{}", s),
            TokenType::Usize(i) => write!(f, "{}", i),
            TokenType::Boolean(b) => write!(f, "{}", b),
            TokenType::Colon => write!(f, ":"),
            TokenType::SemiColon => write!(f, ";"),
            TokenType::Equal => write!(f, "="),
            TokenType::DoubleEqual => write!(f, "=="),
            TokenType::Plus => write!(f, "+"),
            TokenType::Minus => write!(f, "-"),
            TokenType::Multiply => write!(f, "*"),
            TokenType::Divide => write!(f, "/"),
            TokenType::Mod => write!(f, "%"),
            TokenType::NotEqual => write!(f, "!="),
            TokenType::LessThan => write!(f, "<"),
            TokenType::LessThanEqual => write!(f, "<="),
            TokenType::GreaterThan => write!(f, ">"),
            TokenType::GreaterThanEqual => write!(f, ">="),
            TokenType::Bang => write!(f, "!"),
            TokenType::LParen => write!(f, "("),
            TokenType::RParen => write!(f, ")"),
            TokenType::LBrace => write!(f, "{{"),
            TokenType::RBrace => write!(f, "}}"),
            TokenType::Comma => write!(f, ","),
            TokenType::SQuote => write!(f, "'"),
            TokenType::DQuote => write!(f, "\""),
            TokenType::BSlash => write!(f, "\\"),
            TokenType::NewLine => write!(f, "\n"),
            TokenType::Tab => write!(f, "\t"),
            TokenType::Space => write!(f, " "),
            TokenType::KeywordLet => write!(f, "let"),
            TokenType::KeywordUsize => write!(f, "usize"),
            TokenType::KeywordIsize => write!(f, "isize"),
            TokenType::KeywordBool => write!(f, "bool"),
            TokenType::KeywordFun => write!(f, "fun"),
            TokenType::KeywordReturn => write!(f, "return"),
            TokenType::KeywordIf => write!(f, "if"),
            TokenType::KeywordElse => write!(f, "else"),
            TokenType::KeywordVoid => write!(f, "void"),
            TokenType::KeywordChar => write!(f, "chat"),
            TokenType::KeywordWhile => write!(f, "while"),
            TokenType::SymbolReturn => write!(f, "=>"),
            TokenType::String(s) => write!(f, "\"{}\"", s),
            TokenType::LeftShift => write!(f, "<<"),
            TokenType::RightShift => write!(f, ">>"),
            TokenType::Pipe => write!(f, "|"),
            TokenType::Carrot => write!(f, "^"),
            TokenType::Ampersand => write!(f, "&"),
            TokenType::KeywordStr => write!(f, "str"),
        }
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct Token {
    pub type_: TokenType,
    pub pos: (usize, usize),
}

impl Token {
    pub fn new(type_: TokenType, pos: (usize, usize)) -> Self {
        Self { type_, pos }
    }
}
