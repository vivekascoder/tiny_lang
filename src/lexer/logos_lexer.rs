use logos::Logos;

#[derive(Logos, Debug, PartialEq)]
#[logos(skip r"([ \t\r\n]*|//[^\n]*)")]
enum Token {
    #[regex("[a-zA-Z_][a-zA-Z_0-9]*")]
    Ident,
    #[regex(r#""(?:\\.|[^"])*""#)]
    String,
    #[regex("-?[0-9]+")]
    Number,
    #[regex("(true|false)")]
    Boolean,
    #[token(":")]
    Colon,
    #[token(";")]
    SemiColon,

    // Operators
    #[token("=")]
    Equal,
    #[token("==")]
    DoubleEqual,
    #[token("+")]
    Plus,
    #[token("-")]
    Minus,
    #[token("*")]
    Multiply,
    #[token("/")]
    Divide,
    #[token("%")]
    Mod,
    #[token("!=")]
    NotEqual,
    #[token("<")]
    LessThan,
    #[token("<=")]
    LessThanEqual,
    #[token(">")]
    GreaterThan,
    #[token(">=")]
    GreaterThanEqual,
    #[token("!")]
    Bang,
    #[token("(")]
    LParen,
    #[token(")")]
    RParen,
    #[token("{")]
    LBrace, // {
    #[token("}")]
    RBrace, // }
    #[token(",")]
    Comma,
    #[token("'")]
    SQuote, // '
    // #[token("\"")]
    // DQuote, // "
    #[token("\\")]
    BSlash, // \
    #[token("<<")]
    LeftShift, // <<
    #[token(">>")]
    RightShift, // >>
    #[token("&")]
    Ampersand, // &
    #[token("^")]
    Carrot, // ^
    #[token("|")]
    Pipe, // |

    // Keywords
    #[token("let")]
    KeywordLet,
    #[token("usize")]
    KeywordUsize,
    #[token("isize")]
    KeywordIsize,
    #[token("bool")]
    KeywordBool,
    #[token("fun")]
    KeywordFun,
    #[token("return")]
    KeywordReturn,
    #[token("if")]
    KeywordIf,
    #[token("else")]
    KeywordElse,
    #[token("void")]
    KeywordVoid,
    #[token("char")]
    KeywordChar,
    #[token("while")]
    KeywordWhile,

    // Symbol
    #[token("=>")]
    SymbolReturn,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_logos_lexer() {
        let mut lex = Token::lexer("let =");

        loop {
            if let Some(v) = lex.next() {
                println!("> {:?}, {:?}", v, lex.span());
            } else {
                break;
            }
        }
    }

    #[test]
    fn test_function_decl_and_call_logos() {
        let code = r#"
        let a = 445;
        let b = 45;
        let bool_var = false;
        let string_val = "this is a \"string";
        // let char_val = 'a';


        fun sum(a: usize, b: usize) => usize {
            let c = a + b;
            return c;
        }

        sum(a, b);
        "#;
        insta::assert_debug_snapshot!(Token::lexer(code).collect::<Vec<_>>());
    }
}
