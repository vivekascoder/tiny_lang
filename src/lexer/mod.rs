pub mod lexer;
pub mod logos_lexer;
