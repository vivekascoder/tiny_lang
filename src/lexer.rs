use crate::ast::*;
use anyhow::{bail, Result};

#[derive(Debug, Clone, PartialEq)]
pub struct Lexer {
    // To track the current index at the string
    cur: usize,
    source: Vec<char>,
}

/// Parse `let var1: number = 34 + 353;`
impl Lexer {
    pub fn new(s: &str) -> Self {
        Self {
            cur: 0,
            source: s.chars().collect(),
        }
    }

    fn token(&self, type_: TokenType, pos: (usize, usize)) -> Token {
        Token::new(type_, pos)
    }

    fn parse_keyword(&self, s: &str) -> Option<TokenType> {
        match s {
            "let" => Some(TokenType::KeywordLet),
            "usize" => Some(TokenType::KeywordUsize),
            _ => None,
        }
    }

    fn skip_whitespace(&mut self) {
        loop {
            match self.source[self.cur] {
                // Do nothing for whitespace
                ' ' | '\t' | '\n' | '\r' => {
                    self.cur += 1;
                }
                _ => {
                    break;
                }
            }
        }
    }

    pub fn next(&mut self) -> Result<Token> {
        if self.cur == self.source.len() {
            return Ok(self.token(TokenType::EOF, (self.cur, self.cur)));
            // bail!("EOF reached");
        }

        // Skip all the whitespaces till the next token.
        self.skip_whitespace();

        match self.source[self.cur] {
            ';' => {
                self.cur += 1;
                Ok(self.token(TokenType::SemiColon, (self.cur - 1, self.cur)))
            }
            ':' => {
                self.cur += 1;
                Ok(self.token(TokenType::Colon, (self.cur - 1, self.cur)))
            }
            '=' => {
                self.cur += 1;
                Ok(self.token(TokenType::Equal, (self.cur - 1, self.cur)))
            }
            '+' => {
                self.cur += 1;
                Ok(self.token(TokenType::Plus, (self.cur - 1, self.cur)))
            }
            '-' => {
                self.cur += 1;
                Ok(self.token(TokenType::Minus, (self.cur - 1, self.cur)))
            }
            '*' => {
                self.cur += 1;
                Ok(self.token(TokenType::Multiply, (self.cur - 1, self.cur)))
            }
            '/' => {
                self.cur += 1;
                Ok(self.token(TokenType::Divide, (self.cur - 1, self.cur)))
            }
            '!' => {
                self.cur += 1;
                Ok(self.token(TokenType::Bang, (self.cur - 1, self.cur)))
            }
            '%' => {
                self.cur += 1;
                Ok(self.token(TokenType::Mod, (self.cur, self.cur + 1)))
            }
            'a'..='z' | 'A'..='Z' | '_' => {
                // Identifier
                let start = self.cur;

                // Start a loop to parse the identifier.
                loop {
                    match self.source[self.cur] {
                        // TODO: Should break if EOF?
                        'a'..='z' | 'A'..='Z' | '_' | '0'..='9' => {
                            self.cur += 1;
                        }

                        // Encountered something else
                        // Then it could be either identifier or keyword.
                        _ => {
                            let slice = &self.source[start..self.cur];
                            let identifier: String = slice.iter().collect();

                            if let Some(keyword) = self.parse_keyword(&identifier) {
                                return Ok(self.token(keyword, (start, self.cur)));
                            } else {
                                return Ok(self
                                    .token(TokenType::Identifier(identifier), (start, self.cur)));
                            }
                        }
                    }
                }
            }
            '0'..='9' => {
                // Number
                let start = self.cur;

                loop {
                    match self.source[self.cur] {
                        '0'..='9' => {
                            self.cur += 1;
                        }

                        // Encountered any other char.
                        _ => {
                            let s_no = &self.source[start..self.cur];
                            let number: usize = s_no.iter().collect::<String>().parse().unwrap();

                            return Ok(self.token(TokenType::Usize(number), (start, self.cur)));
                        }
                    }
                }
            }

            _ => {
                bail!("Illegal token found: {:?}", self.source[self.cur]);
            }
        }
    }
}

impl Iterator for Lexer {
    type Item = Token;
    fn next(&mut self) -> Option<Self::Item> {
        match self.next() {
            Ok(t) => {
                if t.type_ == TokenType::EOF {
                    None
                } else {
                    Some(t)
                }
            }
            Err(_) => None,
        }
    }
}

#[cfg(test)]
pub mod tests {
    use super::TokenType::*;
    use super::*;
    use std::result::Result::Ok;

    #[test]
    fn test_lexing_let_syntax() {
        let source = "let some_var: usize = 345 + 35353;";
        let mut tokenizer = Lexer::new(source);

        let result = vec![
            Token {
                type_: KeywordLet,
                pos: (0, 3),
            },
            Token {
                type_: TokenType::Identifier(format!("some_var")),

                pos: (4, 12),
            },
            Token {
                type_: TokenType::Colon,
                pos: (12, 13),
            },
            Token {
                type_: KeywordUsize,
                pos: (14, 19),
            },
            Token {
                type_: TokenType::Equal,
                pos: (20, 21),
            },
            Token {
                type_: TokenType::Usize(345),
                pos: (22, 25),
            },
            Token {
                type_: TokenType::Plus,
                pos: (26, 27),
            },
            Token {
                type_: TokenType::Usize(35353),
                pos: (28, 33),
            },
            Token {
                type_: TokenType::SemiColon,
                pos: (33, 34),
            },
            Token {
                type_: TokenType::EOF,
                pos: (34, 34),
            },
        ];
        let mut output = vec![];
        while let Ok(token) = tokenizer.next() {
            output.push(token.clone());
            if token.type_ == TokenType::EOF {
                break;
            }
        }
        assert_eq!(output, result);
    }

    // fn test_parsing_let_without_type() {
    //     let source = "let some_var = 345 + 35353;";
    //     let mut tokenizer = Lexer::new(source);
    //     let tokens = tokenizer.tokenize();
    //     println!("tokens: {:?}", &tokens);

    //     assert_eq!(
    //         tokens,
    //         vec![
    //             Token {
    //                 type_: KeywordLet,
    //                 pos: (0, 3)
    //             },
    //             Token {
    //                 type_: TokenType::Identifier(format!("some_var")),

    //                 pos: (4, 12)
    //             },
    //             Token {
    //                 type_: TokenType::Equal,
    //                 pos: (21, 22)
    //             },
    //             Token {
    //                 type_: TokenType::Usize(345),
    //                 pos: (23, 26)
    //             },
    //             Token {
    //                 type_: TokenType::Plus,
    //                 pos: (27, 28)
    //             },
    //             Token {
    //                 type_: TokenType::Usize(35353),
    //                 pos: (29, 34)
    //             },
    //             Token {
    //                 type_: TokenType::SemiColon,
    //                 pos: (34, 35)
    //             }
    //         ]
    //     );
    // }
}
