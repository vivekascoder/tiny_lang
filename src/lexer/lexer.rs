use std::rc::Rc;

use crate::ast::*;
use anyhow::{bail, Result};
use log::info;

#[derive(Debug, Clone, PartialEq)]
pub struct Lexer {
    // To track the current index at the string
    cur: usize,
    row: usize,
    col: usize,
    module: String,
    source: Vec<char>,
}

/// Parse `let var1: number = 34 + 353;`
impl Lexer {
    pub fn new(module: &str, s: &str) -> Self {
        Self {
            cur: 0,
            row: 1,
            col: 0,
            module: module.to_string(),
            source: s.chars().collect(),
        }
    }

    pub fn module(&self) -> String {
        self.module.clone()
    }

    pub fn get_row(&self) -> usize {
        self.row
    }

    pub fn get_col(&self) -> usize {
        self.col
    }

    fn token(&self, type_: TokenType, pos: (usize, usize)) -> Token {
        Token {
            type_: type_,
            pos: pos,
            row: self.row,
            col: self.col,
        }
    }

    fn parse_keyword(&self, s: &str) -> Option<TokenType> {
        match s {
            "let" => Some(TokenType::KeywordLet),
            "usize" => Some(TokenType::KeywordUsize),
            "isize" => Some(TokenType::KeywordIsize),
            "i8" => Some(TokenType::KeywordI8),
            "fun" => Some(TokenType::KeywordFun),
            "return" => Some(TokenType::KeywordReturn),
            "if" => Some(TokenType::KeywordIf),
            "else" => Some(TokenType::KeywordElse),
            "true" => Some(TokenType::Boolean(true)),
            "false" => Some(TokenType::Boolean(false)),
            "bool" => Some(TokenType::KeywordBool),
            "void" => Some(TokenType::KeywordVoid),
            "char" => Some(TokenType::KeywordChar),
            "while" => Some(TokenType::KeywordWhile),
            "str" => Some(TokenType::KeywordStr),
            "extern" => Some(TokenType::KeywordExtern),
            "struct" => Some(TokenType::KeywordStruct),
            "string" => Some(TokenType::KeywordStr),
            _ => None,
        }
    }

    fn skip_whitespace(&mut self) {
        while !self.eof_reached() {
            match self.current() {
                // Do nothing for whitespace
                ' ' | '\t' | '\r' => {
                    self.bump();
                }
                '\n' => {
                    self.col = 0;
                    self.row += 1;
                    self.bump();
                }
                _ => {
                    break;
                }
            }
        }
    }

    fn current(&self) -> char {
        self.source[self.cur]
    }

    fn bump(&mut self) {
        // if self.eof_reached() {

        // }
        self.cur += 1;
        self.col += 1;
    }

    fn eof_reached(&self) -> bool {
        self.cur == self.source.len()
    }

    pub fn next(&mut self) -> Result<Token> {
        // Skip all the whitespaces till the next token.
        self.skip_whitespace();

        if self.eof_reached() {
            return Ok(self.token(TokenType::EOF, (self.cur, self.cur)));
            // bail!("EOF reached");
        }
        // info!("current token: {:?}", self.current());
        match self.current() {
            ';' => {
                self.bump();
                Ok(self.token(TokenType::SemiColon, (self.cur - 1, self.cur)))
            }
            ':' => {
                self.bump();
                Ok(self.token(TokenType::Colon, (self.cur - 1, self.cur)))
            }

            '.' => {
                let start = self.cur;
                self.bump();
                loop {
                    match self.current() {
                        '.' => {
                            self.bump();
                            continue;
                        }
                        _ => {
                            let periods = &self.source[start..self.cur];
                            if periods.len() == 1 {
                                return Ok(self.token(TokenType::Period, (self.cur - 1, self.cur)));
                            } else if periods.len() == 3 {
                                return Ok(self
                                    .token(TokenType::VariableArg, (self.cur - start, self.cur)));
                            } else {
                                bail!("invalid token `{}`", periods.iter().collect::<String>());
                            }
                        }
                    }
                }
            }

            '=' => {
                self.bump();
                match self.current() {
                    '>' => {
                        self.bump();
                        Ok(self.token(TokenType::SymbolReturn, (self.cur - 2, self.cur)))
                    }
                    '=' => {
                        self.bump();
                        Ok(self.token(TokenType::DoubleEqual, (self.cur - 2, self.cur)))
                    }
                    _ => Ok(self.token(TokenType::Equal, (self.cur - 1, self.cur))),
                }
            }

            '(' => {
                self.bump();
                Ok(self.token(TokenType::LParen, (self.cur - 1, self.cur)))
            }

            '\"' => {
                let start = self.cur;
                loop {
                    self.bump();
                    if self.current() == '\"' {
                        let val: String = (&self.source[start + 1..self.cur]).iter().collect();
                        self.bump();
                        return Ok(self.token(TokenType::String(val.into()), (start, self.cur)));
                    }
                }
            }

            // tokenize whitespace char.
            '\\' => {
                info!("encounterd \\");
                self.bump();
                info!("cur encounterd {}", self.current());
                match self.current() {
                    'n' => {
                        self.bump();
                        return Ok(self.token(TokenType::NewLine, (self.cur - 2, self.cur)));
                    }
                    't' => {
                        self.bump();
                        return Ok(self.token(TokenType::Tab, (self.cur - 2, self.cur)));
                    }
                    // Rust doesn't have `\v` ?
                    // 'v' => Ok(self.token(TokenType::VTab, (self.cur - 2, self.cur))),
                    '\\' => {
                        self.bump();
                        Ok(self.token(TokenType::BSlash, (self.cur - 2, self.cur)))
                    }
                    _ => {
                        bail!("`\\{:?}` is not a valid character.", self.current());
                    }
                }
            }

            '\'' => {
                self.bump();
                Ok(self.token(TokenType::SQuote, (self.cur - 1, self.cur)))
            }

            ')' => {
                self.bump();
                Ok(self.token(TokenType::RParen, (self.cur - 1, self.cur)))
            }

            '&' => {
                self.bump();
                Ok(self.token(TokenType::Ampersand, (self.cur - 1, self.cur)))
            }

            '|' => {
                self.bump();
                Ok(self.token(TokenType::Pipe, (self.cur - 1, self.cur)))
            }

            '^' => {
                self.bump();
                Ok(self.token(TokenType::Carrot, (self.cur - 1, self.cur)))
            }

            '<' => {
                self.bump();
                match self.current() {
                    '<' => {
                        self.bump();
                        Ok(self.token(TokenType::LeftShift, (self.cur - 2, self.cur)))
                    }
                    _ => Ok(self.token(TokenType::LessThan, (self.cur - 1, self.cur))),
                }
            }

            '>' => {
                self.bump();

                match self.current() {
                    '=' => {
                        self.bump();
                        Ok(self.token(TokenType::GreaterThanEqual, (self.cur - 2, self.cur)))
                    }
                    '>' => {
                        self.bump();
                        Ok(self.token(TokenType::RightShift, (self.cur - 2, self.cur)))
                    }
                    _ => Ok(self.token(TokenType::GreaterThan, (self.cur - 1, self.cur))),
                }
            }

            '{' => {
                self.bump();
                Ok(self.token(TokenType::LBrace, (self.cur - 1, self.cur)))
            }

            '}' => {
                self.bump();
                Ok(self.token(TokenType::RBrace, (self.cur - 1, self.cur)))
            }

            ',' => {
                self.bump();
                Ok(self.token(TokenType::Comma, (self.cur - 1, self.cur)))
            }

            '+' => {
                self.bump();
                Ok(self.token(TokenType::Plus, (self.cur - 1, self.cur)))
            }
            '-' => {
                self.bump();
                Ok(self.token(TokenType::Minus, (self.cur - 1, self.cur)))
            }
            '*' => {
                self.bump();
                Ok(self.token(TokenType::Multiply, (self.cur - 1, self.cur)))
            }
            '/' => {
                self.bump();
                match self.current() {
                    // single line comment
                    '/' => loop {
                        self.bump();
                        match self.current() {
                            '\n' => {
                                self.bump();
                                return Ok(self.next()?);
                            }
                            _ => {
                                continue;
                            }
                        };
                    },

                    // multi line comment
                    '*' => loop {
                        match self.current() {
                            '*' => {
                                self.bump();
                                if self.current() == '/' {
                                    self.bump();
                                    return Ok(self.next()?);
                                } else {
                                    continue;
                                }
                            }
                            _ => {
                                self.bump();
                                continue;
                            }
                        };
                    },
                    _ => Ok(self.token(TokenType::Divide, (self.cur - 1, self.cur))),
                }
            }
            '!' => {
                self.bump();
                match self.current() {
                    '=' => {
                        self.bump();
                        Ok(self.token(TokenType::NotEqual, (self.cur - 2, self.cur)))
                    }
                    _ => Ok(self.token(TokenType::Bang, (self.cur - 1, self.cur))),
                }
            }
            '%' => {
                self.bump();
                Ok(self.token(TokenType::Mod, (self.cur, self.cur + 1)))
            }
            'a'..='z' | 'A'..='Z' | '_' => {
                // Identifier
                let start = self.cur;

                // Start a loop to parse the identifier.
                loop {
                    match self.current() {
                        // TODO: Should break if EOF?
                        'a'..='z' | 'A'..='Z' | '_' | '0'..='9' => {
                            self.bump();
                        }

                        // Encountered something else
                        // Then it could be either identifier or keyword.
                        _ => {
                            let slice = &self.source[start..self.cur];
                            let identifier: String = slice.iter().collect();

                            if let Some(keyword) = self.parse_keyword(&identifier) {
                                return Ok(self.token(keyword, (start, self.cur)));
                            } else {
                                return Ok(self.token(
                                    TokenType::Identifier(Rc::from(identifier.as_str())),
                                    (start, self.cur),
                                ));
                            }
                        }
                    }
                }
            }
            '0'..='9' => {
                // Number
                let start = self.cur;

                loop {
                    match self.current() {
                        '0'..='9' => {
                            self.bump();
                        }

                        // Encountered any other char.
                        _ => {
                            let s_no = &self.source[start..self.cur];
                            let number: usize = s_no.iter().collect::<String>().parse().unwrap();

                            return Ok(self.token(TokenType::Usize(number), (start, self.cur)));
                        }
                    }
                }
            }

            _ => {
                bail!("Illegal token found: {:?}", self.current());
            }
        }
    }
}

impl Iterator for Lexer {
    type Item = Token;
    fn next(&mut self) -> Option<Self::Item> {
        match self.next() {
            Ok(t) => {
                if t.type_ == TokenType::EOF {
                    None
                } else {
                    Some(t)
                }
            }
            Err(e) => {
                panic!("expected Token but got error: {:?}", e);
            }
        }
    }
}
