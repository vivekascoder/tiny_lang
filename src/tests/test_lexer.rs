#[cfg(test)]
pub mod tests {
    use crate::ast::*;
    use crate::lexer::Lexer;

    #[test]
    fn test_lexing_let_syntax() {
        let source = "let some_var: usize = 345 + 35353;";
        insta::assert_debug_snapshot!(Lexer::new(source).into_iter().collect::<Vec<Token>>());
    }

    #[test]
    fn test_function_declaration() {
        let source = r#"
        fun calculate_something(a: usize, b: usize) => bool {
            if (a > b) {
                return true;
            } else {
                return false;
            }
        }
        "#;
        insta::assert_debug_snapshot!(Lexer::new(source).into_iter().collect::<Vec<Token>>());
    }
}
