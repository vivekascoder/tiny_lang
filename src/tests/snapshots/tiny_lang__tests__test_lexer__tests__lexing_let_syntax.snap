---
source: src/tests/test_lexer.rs
expression: "Lexer::new(source).into_iter().collect::<Vec<Token>>()"
---
[
    Token {
        type_: KeywordLet,
        pos: (
            0,
            3,
        ),
    },
    Token {
        type_: Identifier(
            "some_var",
        ),
        pos: (
            4,
            12,
        ),
    },
    Token {
        type_: Colon,
        pos: (
            12,
            13,
        ),
    },
    Token {
        type_: KeywordUsize,
        pos: (
            14,
            19,
        ),
    },
    Token {
        type_: Equal,
        pos: (
            20,
            21,
        ),
    },
    Token {
        type_: Usize(
            345,
        ),
        pos: (
            22,
            25,
        ),
    },
    Token {
        type_: Plus,
        pos: (
            26,
            27,
        ),
    },
    Token {
        type_: Usize(
            35353,
        ),
        pos: (
            28,
            33,
        ),
    },
    Token {
        type_: SemiColon,
        pos: (
            33,
            34,
        ),
    },
]
